presto-parser/src/main/java/com/facebook/presto/sql/ExpressionFormatter.java

```diff
@@ -566,6 +570,43 @@ static String formatSortItems(List<SortItem> sortItems, boolean unmangleNames)
                 .iterator());
     }
 
+    static String formatGroupBy(List<GroupingElement> groupingElements)
+    {
+        checkState(groupingElements.stream()
+                .filter(groupingElement -> !(groupingElement instanceof SimpleGroupBy))
+                .count() <= 1, "multiple groupings are not yet supported");
+
+        List<Expression> simpleGroupings = groupingElements.stream()
+                .filter(groupingElement -> groupingElement instanceof SimpleGroupBy)
+                .map(GroupingElement::enumerateGroupingSets)
+                .flatMap(Collection::stream)
+                .flatMap(Collection::stream)
+                .collect(toList());
+
+        if (simpleGroupings.size() > 0) {
+            return Joiner.on(", ").join(simpleGroupings.stream()
+                    .map(ExpressionFormatter::formatExpression)
+                    .iterator());
+        }
+
+        for (GroupingElement groupingElement : groupingElements) {
+            if (groupingElement instanceof GroupingSets) {
+                return "GROUPING SETS (" + Joiner.on(", ").join(
```

Comment:
These nested expressions are hard to read. Either introduce a local variable or use String.format.